src/SubtitlesParser/index.js

```diff
@@ -0,0 +1,199 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Log } from '../LightningSdkPlugins'
+
+export default class SubtitlesParser {
+  // @ params url: subtitle file URL
+  // @ customParser: a customParser to use instead of default parser of the plugin
+  // @ parseOptions.removeSubtitleTextStyles: remove subtitle textstyles possible value true or false
+  // @return parsed subtitles as list of objects
+  // also stores parsed data
+  static fetchAndParseSubs(
+    url,
+    customParser = false,
+    parseOptions = { removeSubtitleTextStyles: true }
+  ) {
+    let _url
+
+    try {
+      _url = new URL(url)
+    } catch (e) {
+      Log.info('Invalid URL')
+      return Promise.reject(new Error('Invalid URL'))
+    }
+    if (!((_url.protocol === 'https:' || _url.protocol === 'http:') && _url.hostname)) {
+      Log.info('Invalid subtitle Url')
+      return Promise.reject(new Error('Invalid URL'))
+    }
+
+    return new Promise((resolve, reject) => {
+      fetch(url)
+        .then(data => data.text())
+        .then(subtitleData => {
+          this.clearAllSubtitles()
+          if (customParser && typeof customParser === 'function') {
+            this._captions = customParser(subtitleData)
+          } else {
+            this._captions = this.parseSubtitles(subtitleData, parseOptions)
+          }
+          if (!this._captions.length) {
+            Log.warn('Invalid subtitles length')
+          }
+          resolve(this._captions)
+        })
+        .catch(error => {
+          Log.error('Fetching subtitles file Failed:', error)
+          this.clearAllSubtitles()
+          reject('Fetching subtitles file Failed')
+        })
+    })
+  }
+
+  // clears stored subtitles data
+  static clearAllSubtitles() {
+    this._captions = []
+    this._lastIndex = 0
+    this._previousCueTimeIndex = 0
+    this._previousCue = ''
+  }
+
+  // get current subtitles
+  // @ currentTime: currentTime in seconds
+  // @return: subtitle text at that currentTime
+  static getSubtitleByTimeIndex(currentTime) {
+    if (typeof currentTime !== 'number' || currentTime === Infinity) {
+      throw new Error('You should pass "currentTime" as a number')
+    }
+
+    if (!Array.isArray(this._captions) || this._captions.length <= 0) {
+      throw new Error('No subtitles available')
+    }
+
+    if (Math.abs(this._previousCueTimeIndex - currentTime) < 0.5) {
+      return this._previousCue
+    }
+
+    if (this._lastIndex > this._captions.length - 1 || !this._lastIndex) {
```

Comment:
I find this place a bit weird for this check, since the `_lastIndex` isn't set in this method (before this point). I feel that this should be wherever the variable is actually updated (so in the `getActiveIndex` method).